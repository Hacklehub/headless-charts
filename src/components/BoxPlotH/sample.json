[
  {
    "name": "Type A",
    "min": 50,
    "firstQuartile": 62,
    "mid": 80,
    "lastQuartile": 90,
    "max": 100
  },
  {
    "name": "Type B",
    "min": 20,
    "firstQuartile": 40,
    "mid": 50,
    "lastQuartile": 60,
    "max": 80
  },
  {
    "name": "Type C",
    "min": 10,
    "firstQuartile": 20,
    "mid": 30,
    "lastQuartile": 40,
    "max": 50
  }
]
