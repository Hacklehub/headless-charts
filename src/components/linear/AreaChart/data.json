[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
