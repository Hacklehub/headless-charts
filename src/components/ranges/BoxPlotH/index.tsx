import { ZoomTransform, zoom } from 'd3-zoom';
import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { max, min } from 'd3-array';
import { pointer, select, selectAll } from 'd3-selection';
import { scaleBand, scaleLinear } from 'd3-scale';

import { ChartProps } from '../../../types';
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { transition } from 'd3-transition';

export interface BoxPlotHProps extends ChartProps {
  classNameData?: string;
  x: {
    minKey: string;
    maxKey: string;
    boxStart: string;
    boxEnd: string;
    midKey: string;
    min?: number;
    max?: number;
    axis?: 'top' | 'bottom';
    axisTicks?: number;
    classNameBoxes?: string;
    classNameLines?: string;
    classNameMap?: any;
  };
  y: {
    key: string;
    axis?: 'left' | 'right';
  };
  tooltip?: {
    className?: string;
    html?: (d: any) => string;
  };
}

const BoxPlotH = ({
  className,
  classNameData,
  data,
  id,
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 20,
  },
  padding = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    bar: 0.2,
  },
  x,
  tooltip,
  y,
  zooming = {
    enabled: false,
  },
}: BoxPlotHProps) => {
  const refreshChart = React.useCallback(() => {
    const svg = select<SVGSVGElement, unknown>(`#${id}`);
    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const g = svg.append('g');

    const xFn = scaleLinear()
      .domain([
        Number.isFinite(x.min) ? x.min : min(data.map((d: any) => d[x.minKey])),
        x.max || max(data.map((d: any) => d[x.maxKey])),
      ])
      .range([margin.left, width - (padding.right || 0) - margin.right]);

    // domainMin is x.min if it exists (including 0), otherwise the min of the data

    const yFn = scaleBand()
      .domain(data.map((d: any) => d[y.key]))
      .range([
        margin.top + (padding.top || 0),
        height - margin.bottom - (padding.bottom || 0),
      ])
      .padding(padding.bar || 0);

    // const clipPath =
    svg
      .append('clipPath')
      .attr('id', 'clip')
      .append('rect')
      .attr('x', margin.left)
      .attr('y', margin.top - (padding.top || 0) - 10)
      .attr('width', width)
      .attr('height', height + (padding.bottom || 0) + 8);

    const dataG = g
      .append('g')
      .attr('class', 'data')
      .attr('clip-path', 'url(#clip)');

    const dotRowsG = dataG
      .selectAll('g')
      .data(data)
      .enter()
      .append('g')
      .on('mouseenter', function (event: MouseEvent, d: any) {
        if (tooltip) {
          tooltipDiv.style('opacity', 1);
          const [bX, bY] = pointer(event, select('body'));
          tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
          tooltipDiv.html(
            tooltip?.html
              ? tooltip.html(d)
              : `min: ${d[x.minKey].toFixed(0)} <br/> range: ${d[
                  x.boxStart
                ].toFixed(0)} to ${d[x.boxEnd].toFixed(0)} <br/> mid: ${d[
                  x.midKey
                ].toFixed(0)} <br/> max: ${d[x.maxKey].toFixed(0)} `
          );
        }
      })
      .on('mousemove', function (event: MouseEvent) {
        const [bX, bY] = pointer(event, select('body'));
        tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
      })
      .on('mouseleave', function () {
        tooltip &&
          tooltipDiv
            .style('opacity', '0')
            .style('left', `0px`)
            .style('top', `0px`);
      });

    transition();

    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.minKey]))
      .attr('x2', (d: any) => xFn(d[x.minKey]))
      .attr('y1', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth() / 2)
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth() / 2)
      .attr('class', (d: any) =>
        mergeTailwindClasses(
          'box-plot-line stroke-current',
          x.classNameMap
            ? x.classNameMap[d[y.key]] || ''
            : x.classNameLines || ''
        )
      )
      .transition()
      .duration(1000)
      .attr('x2', (d: any) => xFn(d[x.maxKey]));

    // Min line
    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.minKey]))
      .attr('x2', (d: any) => xFn(d[x.minKey]))
      .attr('y1', (d: any) => yFn(d[y.key]) || 0)
      .attr('y2', (d: any) => yFn(d[y.key]) || 0)
      .attr('class', (d: any) =>
        mergeTailwindClasses(
          `box-plot-line stroke-current`,
          x.classNameMap
            ? x.classNameMap[d[y.key]] || ''
            : x.classNameLines || ''
        )
      )
      .transition()
      .duration(1000)
      .attr('y1', (d: any) => yFn(d[y.key]) || 0)
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth());

    // Mid line

    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.maxKey]))
      .attr('x2', (d: any) => xFn(d[x.maxKey]))
      .attr('y1', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth())
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth())
      .attr(
        'class',
        mergeTailwindClasses(
          `box-plot-line stroke-current`,
          x.classNameLines || ''
        )
      )
      .transition()
      .duration(1000)
      .attr('y1', (d: any) => yFn(d[y.key]) || 0)
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth());

    dotRowsG
      .append('rect')
      .attr('class', (d: any) =>
        mergeTailwindClasses(
          'box-plot-box stroke-current fill-current',
          d.className,
          x.classNameBoxes || ''
        )
      )
      .attr('clip-path', 'url(#clip)')
      .attr('x', (d: any) => xFn(d[x.boxStart]))
      .attr('y', (d: any) => yFn(d[y.key]) || 0)
      .attr('height', yFn.bandwidth())
      .transition()
      .duration(1000)
      .attr('width', (d: any) => xFn(d[x.boxEnd]) - xFn(d[x.boxStart]));

    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.midKey]))
      .attr('x2', (d: any) => xFn(d[x.midKey]))
      .attr('y1', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth() / 2)
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth() / 2)
      .attr(
        'class',
        mergeTailwindClasses(`box-plot-line stroke-current`, classNameData)
      )
      .transition()
      .duration(1000)
      .attr('y1', (d: any) => yFn(d[y.key]) || 0)
      .attr('y2', (d: any) => (yFn(d[y.key]) || 0) + yFn.bandwidth());

    const tooltipDiv = select('body')
      .append('div')
      .attr('id', 'tooltip')
      .style('position', 'absolute')
      .style('opacity', '0')
      .attr('class', `tooltip ${(tooltip && tooltip.className) || ''}`);

    const yAxis = y.axis === 'right' ? axisRight(yFn) : axisLeft(yFn);

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${y.axis === 'right' ? margin.left + width : margin.left},0)`
      );

    yAxisG.call(yAxis);

    padding.bottom &&
      yAxisG
        .append('line')
        .attr('x1', 0)
        .attr('x2', 0)
        .attr('y1', margin.top + height - padding.bottom)
        .attr('y2', margin.top + height)
        .attr('stroke', 'currentColor');

    const xAxis =
      x.axis === 'top'
        ? axisTop(xFn).ticks(x.axisTicks || 5)
        : axisBottom(xFn).ticks(x.axisTicks || 5);

    const xAxisG = g.append('g').attr('class', 'axis--x axis ');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top : height - margin.bottom
        })`
      )
      .call(xAxis);
    if (zooming?.enabled) {
      const zoomed = ({ transform }: { transform: ZoomTransform }) => {
        // transform g across only x axis
        dotRowsG.attr(
          'transform',
          `translate(${transform.x},0)scale(${transform.k},1)`
        );
        xAxisG.call(xAxis.scale(transform.rescaleX(xFn)));
      };

      const zoomFn = zoom<SVGSVGElement, unknown>()
        .scaleExtent([zooming.min || 1, zooming.max || 2])
        .translateExtent([
          [0, 0],
          [width, height],
        ])
        .on('zoom', zoomed);

      svg.call(zoomFn);
    }
  }, [data, id, margin, padding, x, y, tooltip, classNameData, zooming]);

  React.useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data, refreshChart]);

  return (
    <>
      <svg
        id={id}
        className={mergeTailwindClasses(
          defaultChartClassNames,
          className || ''
        )}
      />
    </>
  );
};

export default BoxPlotH;
