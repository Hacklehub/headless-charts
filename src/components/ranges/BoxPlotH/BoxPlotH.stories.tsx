/* eslint-disable @typescript-eslint/no-explicit-any */
import BoxPlotH from '.';
import { Meta } from '@storybook/react';
import data from '../sample.json';

export default {
  title: 'Ranges/BoxPlotH',
  component: BoxPlotH,
  tags: ['autodocs'],
} as Meta;

/** Default BoxPlot (Grouped). */
export const Default = {
  args: {
    data,
    id: 'box-plot-h-default',
    x: {
      minKey: 'min',
      maxKey: 'max',
      midKey: 'mid',
      boxStart: 'firstQuartile',
      boxEnd: 'lastQuartile',
      min: 0,
    },
    y: { key: 'name' },
  },
};

/** BoxPlot (Grouped) with custom colors. */
export const CustomColors = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-colors',
    x: {
      ...Default.args.x,
      classNameBoxes: 'text-blue-500 opacity-100',
    },
  },
};

export const CustomColorMap = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-color-map',
    x: {
      ...Default.args.x,
    },
    data: data.map((d: any, idx: number) => ({
      ...d,
      className: [
        'text-red-900',
        'text-green-900',
        'text-yellow-500',
        'text-orange-500',
      ][idx],
    })),
  },
};

export const DefaultTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-default-tooltip',
    tooltip: {},
  },
};

export const CustomStyleTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-tooltip',
    tooltip: {
      className: 'bg-gray-100 text-gray-900',
    },
  },
};

export const CustomHtmlTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-html-tooltip',
    tooltip: {
      html: (d: any) => {
        return `<div class="bg-gray-100 text-gray-900 p-2 rounded">
          <div class="text-lg font-bold">${d.name}</div>
          <div class="text-sm">${d.name} ranges between ${d.min} and ${d.max}, with an median of ${d.mid}. </div>
          <div class="text-sm">The first and last quartile values are at ${d.firstQuartile} and ${d.lastQuartile}</div>
        </div>`;
      },
    },
  },
};
