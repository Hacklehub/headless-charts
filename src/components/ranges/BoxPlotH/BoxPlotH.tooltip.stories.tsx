/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import BoxPlotH from '.';
import data from '../sample.json';
import xValues from './sample.json';

const meta: Meta<typeof BoxPlotH> = {
  title: 'Ranges/BoxPlotH/Tooltips',
  component: BoxPlotH,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof BoxPlotH>;

export const DefaultTooltip:Story = {
  args: {
    data,
    x: xValues,
    y: { key: 'name' },
    id: 'box-plot-h-default-tooltip',
    tooltip: {},
  },
};

export const CustomStyleTooltip: Story = {
  args: {
    data,
    x: xValues,
    y: { key: 'name' },
    id: 'box-plot-h-custom-tooltip',
    tooltip: {
      className: 'bg-gray-100 text-gray-900',
    },
  },
};

export const CustomHtmlTooltip: Story = {
  args: {
    data,
    x: xValues,
    y: { key: 'name' },
    id: 'box-plot-h-custom-html-tooltip',
    tooltip: {
      html: (d: any) => {
        return `<div class="bg-gray-100 text-gray-900 p-2 rounded">
            <div class="text-lg font-bold">${d.name}</div>
            <div class="text-sm">${d.name} ranges between ${d.min} and ${d.max}, with an median of ${d.mid}. </div>
            <div class="text-sm">The first and last quartile values are at ${d.firstQuartile} and ${d.lastQuartile}</div>
          </div>`;
      },
    },
  },
};
