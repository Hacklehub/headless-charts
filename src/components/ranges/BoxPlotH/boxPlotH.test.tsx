import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import BoxPlotH from '.';
import data from '../sample.json';

describe('boxPlotH', () => {
  it('renders BoxPlotH component for snapshot test', () => {
    render(
      <BoxPlotH
        data={data}
        id='boxPlotH'
        x={{
          minKey: 'min',
          maxKey: 'max',
          midKey: 'mid',
          boxStart: 'firstQuartile',
          boxEnd: 'lastQuartile',
          min: 0,
        }}
        y={{ key: 'name' }}
      />
    );
    const boxPlotH = screen.getByTestId('boxPlotH');
    expect(boxPlotH).toMatchSnapshot();
  });
  it('renders BoxPlotH component', () => {
    render(
      <BoxPlotH
        data={data}
        id='boxPlotH'
        x={{
          minKey: 'min',
          maxKey: 'max',
          midKey: 'mid',
          boxStart: 'firstQuartile',
          boxEnd: 'lastQuartile',
          min: 0,
        }}
        y={{ key: 'name' }}
      />
    );
    const boxPlotH = screen.getByTestId('boxPlotH');
    expect(boxPlotH).toBeInTheDocument();
  });
});
