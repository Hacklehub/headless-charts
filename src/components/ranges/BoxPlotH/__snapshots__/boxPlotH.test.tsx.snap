// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`boxPlotH > renders BoxPlotH component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="boxPlotH"
  id="boxPlotH"
>
  <g>
    <g
      class="data"
      clip-path="url(#clip)"
    >
      <g>
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="10"
          x2="10"
          y1="28.57142857142857"
          y2="28.57142857142857"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="10"
          x2="10"
          y1="20.95238095238095"
          y2="20.95238095238095"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="-20"
          x2="-20"
          y1="36.19047619047619"
          y2="36.19047619047619"
        />
        <rect
          class="box-plot-box fill-current stroke-current"
          clip-path="url(#clip)"
          height="15.238095238095239"
          x="2.799999999999999"
          y="20.95238095238095"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="-8.000000000000002"
          x2="-8.000000000000002"
          y1="28.57142857142857"
          y2="28.57142857142857"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="40"
          x2="40"
          y1="9.523809523809526"
          y2="9.523809523809526"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="40"
          x2="40"
          y1="1.904761904761905"
          y2="1.904761904761905"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="16"
          x2="16"
          y1="17.142857142857146"
          y2="17.142857142857146"
        />
        <rect
          class="box-plot-box fill-current stroke-current"
          clip-path="url(#clip)"
          height="15.238095238095239"
          x="34"
          y="1.904761904761905"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="28"
          x2="28"
          y1="9.523809523809526"
          y2="9.523809523809526"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="28"
          x2="28"
          y1="-9.523809523809522"
          y2="-9.523809523809522"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="28"
          x2="28"
          y1="-17.142857142857142"
          y2="-17.142857142857142"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="-8.000000000000002"
          x2="-8.000000000000002"
          y1="-1.9047619047619033"
          y2="-1.9047619047619033"
        />
        <rect
          class="box-plot-box fill-current stroke-current"
          clip-path="url(#clip)"
          height="15.238095238095239"
          x="16"
          y="-17.142857142857142"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="10"
          x2="10"
          y1="-9.523809523809522"
          y2="-9.523809523809522"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="34"
          x2="34"
          y1="-28.57142857142857"
          y2="-28.57142857142857"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="34"
          x2="34"
          y1="-36.19047619047619"
          y2="-36.19047619047619"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="10"
          x2="10"
          y1="-20.95238095238095"
          y2="-20.95238095238095"
        />
        <rect
          class="box-plot-box fill-current stroke-current"
          clip-path="url(#clip)"
          height="15.238095238095239"
          x="28"
          y="-36.19047619047619"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="22"
          x2="22"
          y1="-28.57142857142857"
          y2="-28.57142857142857"
        />
      </g>
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,40.5H0.5V-39.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,28.57142857142857)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type A
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,9.523809523809526)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-9.523809523809522)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type C
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-28.57142857142857)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type D
        </text>
      </g>
    </g>
    <g
      class="axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H-19.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(28.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(16.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(4.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-7.500000000000002,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-19.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100
        </text>
      </g>
    </g>
  </g>
  <clippath
    id="clip"
  >
    <rect
      height="8"
      width="0"
      x="40"
      y="30"
    />
  </clippath>
</svg>
`;
